helix-term/src/ui/editor.rs

```diff
@@ -697,6 +738,242 @@ impl EditorView {
+        }
+
+        let mut parent = tree
+            .root_node()
+            .descendant_for_byte_range(top_first_byte, top_first_byte)
+            .and_then(|n| n.parent());
+
+        // context is list of numbers of lines that should be rendered in the LSP context
+        let mut context: Vec<StickyNode> = Vec::new();
+
+        while let Some(node) = parent {
+            // skip all the nodes that aren't in the scope of the cursor
+            if !cursor_nodes
+                .iter()
+                .any(|cursor_node| cursor_node.start_byte() == node.start_byte())
+            {
+                parent = node.parent();
+                continue;
+            }
+
+            let line = text.byte_to_line(node.start_byte());
+
+            // if parent of previous node is still on the same line, use the parent node
+            if let Some(prev_line) = context.last() {
+                if prev_line.line_nr == line {
+                    context.pop();
+                }
+            }
+
+            let mut cursor = QueryCursor::new();
+            let query = &context_nodes.query;
+            let query_nodes = cursor.matches(query, node, RopeProvider(text));
+
+            let query_ranges = query_nodes
+                .flat_map(|qnode| {
+                    qnode
+                        .captures
+                        .iter()
+                        .map(|capture| capture.node.byte_range())
+                })
+                .collect::<Vec<std::ops::Range<usize>>>();
+
+            if query_ranges
+                .iter()
+                .any(|query_range| query_range.contains(&node.start_byte()))
+            {
+                context.push(StickyNode {
+                    visual_line: 0, // with sorting it will be done
+                    line_nr: line,
+                    indicator: None,
+                    top_first_byte,
+                });
+            }
+
+            parent = node.parent();
+        }
+
+        // context should be filled by now
+        if context.is_empty() {
+            return None;
+        }
+
+        let max_lines = config.sticky_context.max_lines;
+        let max_nodes_amount = if max_lines == 0 {
+            viewport.height as usize / 3
+        } else {
+            max_lines.min(viewport.height) as usize
+        };
+
+        // we render from top most (last in the list)
+        context = context
+            .into_iter()
+            // only take the nodes until 1 / 3 of the viewport is reached or the maximum amount of sticky nodes
+            .take(max_nodes_amount)
+            .rev()
+            .enumerate()
+            .take_while(|(i, _)| *i + 1 != visual_cursor_pos as usize) // also only nodes that don't overlap with the visual cursor position
+            .map(|(i, node)| {
+                let mut new_node = node;
+                new_node.visual_line = i as u16;
+                new_node
+            })
+            .collect();
```

Comment:
This will take `max_node_amount` and then remove elements again if they overlap, is that intended ?

The version below avoids one `Vec` allocation but I'm not convinced it would be faster, how big are you expecting `context` to be most of the time ?

```rust
context.reverse();
if context.len() > max_node_amount {
    // only take the nodes until 1 / 3 of the viewport is reached or the maximum amount of sticky nodes
    context.drain(max_node_amount..);
}
let mut idx = 0;
let mut overlap_reached = 1 == visual_cursor_pos;
context.retain_mut(|node| {
    if overlap_reached { return false; }
    overlap_reached = idx + 1 == visual_cursor_pos;
    node.visual_line = idx;
    idx += 1;
    true
});
```